proof/refine/ARM/Tcb_R.thy

```diff
@@ -3097,100 +3144,176 @@ lemma tc_corres_sched:
+                         valid_release_queue'_def inQ_def
+                   elim: ps_clear_domE valid_objsE'[where obj="KOSchedContext sc"] split: if_splits)
+  done
+
+lemma schedContextBindTCB_invs':
+  "\<lbrace>\<lambda>s. invs' s \<and> ex_nonz_cap_to' tcbPtr s \<and> ex_nonz_cap_to' scPtr s \<and>
+        (ksSchedulerAction s = ResumeCurrentThread \<longrightarrow> ksCurThread s \<noteq> tcbPtr) \<and>
+        bound_sc_tcb_at' (\<lambda>sc. sc = None) tcbPtr s \<and> obj_at' (\<lambda>sc. scTCB sc = None) scPtr s\<rbrace>
+   schedContextBindTCB scPtr tcbPtr
+   \<lbrace>\<lambda>_. invs'\<rbrace>"
+  apply (simp add: schedContextBindTCB_def)
+  apply (subst bind_assoc[symmetric, where m="threadSet _ _"])
+  apply (rule hoare_seq_ext)+
+      apply wpsimp
+     apply (wpsimp wp: isSchedulable_wp)
+    apply (clarsimp simp: isSchedulable_bool_runnableE)
+    apply (wp (once) hoare_drop_imps, wp (once) hoare_drop_imps, wp hoare_vcg_imp_lift')
+    apply (wp (once) hoare_drop_imps, wp (once) hoare_drop_imps, wp hoare_vcg_imp_lift')
+   apply (clarsimp cong: conj_cong)
+   apply (rule hoare_vcg_conj_lift[OF schedContextBindTCB_invs'_helper])
+   apply (wp hoare_vcg_imp_lift')
+  apply wp
+  apply (clarsimp simp: obj_at'_def)
+  done
+
+lemma threadSetPriority_bound_sc_tcb_at' [wp]:
+  "threadSetPriority tptr prio \<lbrace>\<lambda>s. Q (bound_sc_tcb_at' P t s)\<rbrace>"
+  by (wpsimp wp: threadSet_pred_tcb_no_state simp: threadSetPriority_def)
+
+crunches setPriority
+  for sc_tcb_sc_at'[wp]: "\<lambda>s. Q (obj_at' (\<lambda>sc. P (scTCB sc)) p s)"
+  and bound_sc_tcb_at'[wp]: "\<lambda>s. Q (bound_sc_tcb_at' P t s)"
+  and ksSchedulerAction[wp]: "\<lambda>s. P (ksSchedulerAction s)"
+  and ksCurThread[wp]: "\<lambda>s. P (ksCurThread s)"
+  (wp: crunch_wps)
+
+crunches setMCPriority
+  for sc_tcb_sc_at'[wp]: "\<lambda>s. Q (obj_at' (\<lambda>sc. P (scTCB sc)) p s)"
+  and bound_sc_tcb_at'[wp]: "\<lambda>s. Q (bound_sc_tcb_at' P t s)"
+  and st_tcb_at'[wp]: "\<lambda>s. Q (st_tcb_at' P t s)"
+  and ksSchedulerAction[wp]: "\<lambda>s. P (ksSchedulerAction s)"
+  and ksReadyQueues[wp]: "\<lambda>s. P (ksReadyQueues s)"
+  (wp: crunch_wps threadSet_pred_tcb_no_state)
+
+crunches finaliseCap, capSwapForDelete
+  for ksCurThread[wp]: "\<lambda>s. P (ksCurThread s)"
+  (simp: crunch_simps wp: crunch_wps getObject_inv loadObject_default_inv cteDelete_preservation)
 
+crunches cteInsert, emptySlot, cancelAllIPC
+  for sc_tcb_sc_at'[wp]: "\<lambda>s. Q (obj_at' (\<lambda>sc. P (scTCB sc)) p s)"
+  (wp: crunch_wps)
+
+lemma installTCBCap_sc_tcb_sc_at:
+  "\<lbrace>\<lambda>s. Q (obj_at' (\<lambda>sc. P (scTCB sc)) p s) \<and> invs' s \<and> tcb_at' target s \<and>
+        cte_wp_at' (isValidFaultHandler \<circ> cteCap) (cte_map (target, tcb_cnode_index 3)) s\<rbrace>
+   installTCBCap target slot 3 slot_opt
+   \<lbrace>\<lambda>_ s. Q (obj_at' (\<lambda>sc. P (scTCB sc)) p s)\<rbrace>"
```

Comment:
These lemmas are specifically for the threadFaultHandlerSlot (slot 3). Maybe that should be shown in the name as well?